src/dataframe/dataframe.jl

```diff
@@ -1153,13 +1166,49 @@ function Base.append!(df1::DataFrame, df2::AbstractDataFrame; cols::Symbol=:sete
     try
         for (j, n) in enumerate(_names(df1))
             current_col += 1
-            append!(df1[!, j], df2[!, n])
+            if n in _names(df2)
+                df1_c = df1[!, j]
+                df2_c = df2[!, n]
+                S = eltype(df2_c)
+                T = eltype(df1_c)
+                V = promote_type(S, T)
+                if S <: T || V <: T || !promote
+                    append!(df1_c, df2_c)
+                else
+                    newcol = Tables.allocatecolumn(V, targetrows)
```

Comment:
This might preserve more properties of the input column:
```suggestion
                    newcol = similar(df1_c, V, targetrows)
```

Reply:
Yeah - this is always tricky.